src-tests/test-rpc.lisp

```diff
@@ -7,87 +7,110 @@
 
 (defparameter *expected-response* "test-response")
 
-(defun test-method (&key (sleep 1 sleep-p))
-  (when sleep-p
+(defun test-method (&key (sleep 1 slee-p))
+  (when slee-p
     (sleep sleep))
   "test-response")
 
+(defmacro with-unique-rpc-address ((addr) &body body)
+  `(let ((,addr (concatenate 'string "inproc://" (format nil "~a" (uuid:make-v4-uuid)))))
+     ,@body))
 
 (deftest test-client-server-dialogue ()
-  (let* ((server-function
-           (lambda ()
-             (let ((dt (rpcq:make-dispatch-table)))
-               (rpcq:dispatch-table-add-handler dt 'test-method)
-               (rpcq:start-server :dispatch-table dt
-                                  :listen-addresses '("inproc://RPCQ-test")))))
-         (server-thread (bt:make-thread server-function)))
-    (sleep 1)
-    (unwind-protect
-      ;; hook up the client
-      (rpcq:with-rpc-client (client "inproc://RPCQ-test")
-        ;; send a communique
-        (let ((server-response (rpcq:rpc-call client "test-method")))
-          (is (string= *expected-response* server-response))))
-      ;; kill the server thread
-      (bt:destroy-thread server-thread))))
+  (with-unique-rpc-address (addr)
+    (let* ((server-function
+             (lambda ()
+               (let ((dt (rpcq:make-dispatch-table)))
+                 (rpcq:dispatch-table-add-handler dt 'test-method)
+                 (rpcq:start-server :dispatch-table dt
+                                    :listen-addresses (list addr)))))
+           (server-thread (bt:make-thread server-function)))
+      (sleep 1)
+      (unwind-protect
+           ;; hook up the client
+           (rpcq:with-rpc-client (client addr)
+             ;; send a communique
+             (let ((server-response (rpcq:rpc-call client "test-method")))
+               (is (string= *expected-response* server-response))))
+        ;; kill the server thread
+        #+ccl
+        (loop :while (bt:thread-alive-p server-thread)
```

Comment:
`bt:join-thread` no worky?

https://trac.common-lisp.net/bordeaux-threads/wiki/ApiDocumentation#join-threadthread

Reply:
Alas, no. I don't quite understand the threading logic that RPCQ employs, so I cannot say why `join-thread` doesn't work.

Reply:
Huh. That is strange. I think this is fine to get CCL support unblocked, but I would like to understand why `join-thread` doesn't work, if only for my curiosity. I'll take a look at this later today if I have time. Do I need to do anything special to reproduce, or just run these tests under CCL?

Reply:
Turns out loading `pzmq` fails for me on CCL/Linux, which means I also can't load `rpcq`.

```
Error while trying to load definition for system pzmq from
pathname
/home/ma/opt/quicklisp/dists/quicklisp/software/pzmq-20171019-git/pzmq.asd:

   While compiling BUNDLE-PATHNAME-TYPE :
In the form (#1=(UIOP/OS:OS-UNIX-P)
             "o"), #1# is not a symbol or lambda expression.

```

Reply:
You wouldn't know it from reading the bordeaux-threads docs, but I guess calling `bt:destroy-thread` followed by `bt:join-thread` is not valid on some implementations. SBCL, for example, signals an error because you're attempting to join to a thread that was aborted (as opposed to exiting normally) without providing a default return value. The CCL docs claim that CCL will signal error as well, but it seems that is not actually the case. So I guess `join-thread` is not a great option here.

I think the ideal solution is to update `rpcq:start-server` to allow for graceful termination of the worker threads, assuming that is possible, rather than relying on `bt:destroy-thread` which is a harsh way to terminate, doesn't guarantee cleanup forms get executed, etc. It might be possible to kill two birds with one stone and add a non-blocking `start-server` option while we're at it (#61).

But I think this simple `thread-alive-p` loop is fine for this PR to unblock CCL support, as long as it works reliably and doesn't result in test flakiness.

If no one else is relishing the thought of adding graceful shutdown to `rpcq:start-server`, then I'd be happy to file an issue and take a look at some point.

As for what's going wrong on CCL, I have a vague suspicion that it might be a bad interaction between the way `bt:destroy-thread` is implemented on CCL and `pzmq:*restart-interrupted-calls*`, which is `t` by default and causes `pzmq` to retry blocking calls that get interrupted by `EINTR`. But that is just a hunch and requires more digging to get to the bottom of it. It's probably not worth the effort to debug it as long as adding graceful shutdown is workable.

Reply:
Filing a cross-referenced issue from #61 about graceful shutdown sounds like the best plan. I have no idea how to get the `pzmq:device` call to terminate, and I imagine it's tricky / might require a more manual implementation of that ZMQ pattern. On the other hand, maybe it doesn't matter too much if we should do the ultimate shutdown forcefully, provided the workers first empty themselves gracefully.

Reply:
Ty for chasing this down some!

Reply:
Sounds good to me. I'll file an issue and link it here.

Reply:
#75

Reply:
P.S. @notmgsk if you feel like tackling #75 as part of this PR (or after), be my guest. Don't want to steal your :zap: thunder!  Alternatively, if you have better things to do, feel free to leave it to me.

Reply:
Let's get this merged and talk about that PR after.